src/Orchard.Web/Modules/Orchard.MessageBus/Services/DistributedShellStarter.cs

```diff
@@ -33,12 +38,21 @@ public class DistributedShellStarter : IDistributedShellStarter, IOrchardShellEv
 
                         // todo: this doesn't work as the new tenants list is lost right after
                         var shellSettingsManagerEventHandler = scope.Resolve<IShellSettingsManagerEventHandler>();
-                        shellSettingsManagerEventHandler.Saved(shellSettings);
-
                         var orchardHost = scope.Resolve<IOrchardHost>() as DefaultOrchardHost;
-                        if(orchardHost != null) {
-                            var startUpdatedShellsMethod = typeof(DefaultOrchardHost).GetMethod("StartUpdatedShells", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
-                            startUpdatedShellsMethod.Invoke(orchardHost, null);
+
+                        // We only want a single thread setting the "IsStarting" flag
+                        lock (locker) { 
```

Comment:
Do you really mean `lock` where the next thread will also execute this code, or `Monitor.TryEnter` where the first one executes and not the others?